### PR TITLE
LWS v2.4.2, mbedTLS v2.8.0, Websocket  SSL support

### LINKED ISSUE #17587: WebSocketClient connect_to_url does not connect
**Godot version:**

201d2d7226ba4d385ac26e3cfaada180c2a7dacb

**OS/device including version:**

Windows 10 Pro v1709

**Issue description:**

Trying to test out WebSocketClient, nothing happens after calling the connect_to_url. 

**Minimal reproduction project:**

```
extends Node

var ws = null

func _ready():
	ws = WebSocketClient.new()
	ws.connect("connection_established", self, "_connection_established")
	ws.connect("connection_closed", self, "_connection_closed")
	ws.connect("connection_error", self, "_connection_error")
	
	var url = "wss://echo.websocket.org"
	print("Connecting to " + url)
	ws.connect_to_url(url)
	
func _connection_established(protocol):
	print("Connection established with protocol: ", protocol)
	
func _connection_closed():
	print("Connection closed")

func _connection_error():
	print("Connection error")
```

### PR DESCRIPTION
**Pull request by Faless**
- Update `libwebsockets` to version `2.4.2`
  - plus update the README with details of patches
- Update `mbedTLS` to version `2.8.0`
  - plus update the README with details of patches
- Native `WebsocketClient` now supports SSL with (optional) verification in native builds using the project defined SSL certs.
   (in `HTML5` version SSL verification is mandatory, and depends on browser trusted certificates).
- (minor) Move the function to read the certificate file to `StreamPeerSSL` making it reusable.
  Ideally, in the future, we might add per peer certs (even loadable via GDscript), but that requires a bit of discussion so I left it as it was.

Closes #17587

### CODE REVIEW

File: thirdparty/lws/mbedtls_wrapper/include/platform/ssl_port.h [hunk 1]
```diff
@@ -24,7 +24,12 @@
 #include "esp_log.h"
 */
 #include "string.h"
+
+#if defined(__APPLE__) || defined(__FreeBSD__) || defined(__OpenBSD__)
```
Comment (akien-mga) [inline thirdparty/lws/mbedtls_wrapper/include/platform/ssl_port.h]: I guess you could merge this one in its parent commit, which undid the fix. The parent commit also has some Godot-specific changes, so you can just mark those as Godot-specific until they're actually replaced by an upstream release.

File: thirdparty/README.md [hunk 2]
```diff
@@ -246,6 +246,7 @@ File extracted from upstream source:
 - Also copy `win32helpers/` from `win32port/`
 - `mbedtls_wrapper/include/platform/ssl_port.h` has a small change to check for OSX and FreeBSD (missing `malloc.h`).
   The bug is fixed in upstream master via `LWS_HAVE_MALLOC_H`, but not in the 2.4.1 branch (as the file structure has changed).
+- You might need to apply the patch in `thirdparty/lws/mbedtls_verify.diff` (port of PR 1215) to future `2.4.x` releases if it does not get cherry picked.
```
Comment (akien-mga) [inline thirdparty/README.md]: This commit could be squashed into 1ac806cb8fe2f9332ecce5ce2f7ddf389bebb472.

File: thirdparty/mbedtls/library/net_sockets.c [hunk 3]
```diff
@@ -49,7 +49,7 @@
 #undef _WIN32_WINNT
 #endif
 /* Enables getaddrinfo() & Co */
-#define _WIN32_WINNT 0x0501
+#define _WIN32_WINNT 0x0601
```
Comment (akien-mga) [inline thirdparty/mbedtls/library/net_sockets.c]: Any chance to get this fixed upstream? It's bogus of them to just override `_WIN32_WINNT` with whatever they need as minimum. They should do something like:
```
#if !defined(_WIN32_WINNT) || (_WIN32_WINNT < 0x0501)
/* Enables getaddrinfo() & Co */
#define _WIN32_WINNT 0x0501
#endif
```

**Review by akien-mga — APPROVED** [review]

**Comment by akien-mga** [comment]
I wonder why GitHub gets confused about the ordering of your commits most of the time. `git log` gets it right, but GH is messed up.. (doesn't matter much, but it's a bit random when reviewing in the order shown by GH).

**Comment by Faless** [comment]
> I wonder why GitHub gets confused about the ordering of your commits most of the time. git log gets it right, but GH is messed up..

Yeah, GitHub doesn't like heavy rebasing. It orders commit by date, not the actual commit order in git

**Comment by Faless** [comment]
Squashed and updated as per PR review, should be ready to merge when CI finishes.

**Comment by akien-mga** [comment]
Thanks!

**Pull request merged**